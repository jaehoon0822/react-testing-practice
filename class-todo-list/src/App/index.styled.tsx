import styled from '@emotion/styled';

export const Container = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  min-height: 100vh;
  border-radius: 4px;
`;

export const Contents = styled.div`
  display: flex;
  flex-direction: column;
  min-width: 400px;
  padding: 1rem;
  box-shadow: 10px 10px 20px #bdbdbd;
`;

export const ToDoListContainer = styled.div`
  width: 100%;
  height: 400px;
  margin-bottom: 1rem;
  border: 1px solid #bdbdbd;
  overflow-y: scroll;
`;

export const InputContainer = styled.div`
  display: flex;
`