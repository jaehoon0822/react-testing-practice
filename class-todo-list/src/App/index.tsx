import { ChangeEvent, Component } from 'react';
import { Props, State } from './index.types';
import { Container, Contents, InputContainer, ToDoListContainer } from './index.styled';
import Input from 'Components/Input';
import { Button } from 'Components/Button';
import ToDoItem from 'Components/ToDoItem';
import { nanoid } from 'nanoid';

export default class App extends Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      toDo: '',
      toDoList: [],
    };
  }

  private onChangeToDo = (e: ChangeEvent<HTMLInputElement>) => {
    this.setState({
      toDo: e.target.value,
    });
  };

  private onClickDelete = (idx: number) => {
    const newToDoList = this.state.toDoList;
    newToDoList.splice(idx, 1);
    this.setState({
      toDoList: newToDoList,
    });
  };

  private onClickAdd = () => {
    if (this.state.toDo !== '') {
      this.setState({
        toDoList: [
          {
            key: nanoid(),
            label: this.state.toDo,
          },
          ...this.state.toDoList,
        ],
        toDo: '',
      });
    }
  };

  render() {
    const { toDo, toDoList } = this.state;
    return (
      <Container>
        <Contents>
          <ToDoListContainer>
            {toDoList.map((item, idx) => (
              <ToDoItem
                key={item.key}
                label={item.label}
                onClickDelete={() => this.onClickDelete(idx)}
              />
            ))}
          </ToDoListContainer>
          <InputContainer>
            <Input
              label="할일"
              name="toDo"
              LabelDisplay="none"
              value={toDo}
              onChange={this.onChangeToDo}
            />
            <Button label="추가" onClick={this.onClickAdd} />
          </InputContainer>
        </Contents>
      </Container>
    );
  }
}
