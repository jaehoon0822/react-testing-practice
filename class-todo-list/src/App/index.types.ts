import { IToDoItem } from 'Components/ToDoItem/index.types';

export interface State {
  readonly toDo: string;
  readonly toDoList: IToDoItem[];
}

export interface Props {}
