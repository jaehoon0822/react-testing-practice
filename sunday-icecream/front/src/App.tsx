import styles from "./App.module.scss";
function App() {
  return (
    <div className={styles.container}>
      <button className={styles.btn}>hi</button>
    </div>
  );
}

export default App;
