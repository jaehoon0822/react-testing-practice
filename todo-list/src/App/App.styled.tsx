import styled from '@emotion/styled';

export const Container = styled.div`
  min-height: 100vh;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
`;

export const Contents = styled.div`
  display: flex;
  background-color: #fff;
  flex-direction: column;
  padding: 20px;
  border-radius: 8px;
  box-shadow: 5px 5px 10px 10px rgba(0, 0, 0, 0.2);
`;

export const InputContainer = styled.div`
  display: flex;
`;

export const ToDoListContainer = styled.div`
  padding-bottom: 1rem;
  min-width: 350px;
  height: 400px;
  overflow-y: scroll;
  border: 1px solid #bdbdbd;
  margin-bottom: 20px;
`
