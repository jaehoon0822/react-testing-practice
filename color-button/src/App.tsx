import React, { useMemo, useState } from "react";
import "./App.css";

export function replaceCamelWithSpaces(colorName: string) {
  return colorName.replace(/\B([A-Z])\B/g, " $1");
}

function App() {
  const [isDisabled, setIsDisabled] = useState<boolean>(false);
  const [buttonColor, setButtonColor] = useState<String>("blue");

  const buttonStyle: React.CSSProperties = useMemo(() => {
    const bgColor = buttonColor === "red" ? "blue" : "red";

    return {
      backgroundColor: isDisabled ? "gray" : bgColor,
      color: "white",
    };
  }, [buttonColor, isDisabled]);

  const onClickCheckbox = (e: React.ChangeEvent<HTMLInputElement>) => {
    setIsDisabled((_prev) => e.target.checked);
  };

  const onClickButton = () => {
    setButtonColor((prev) => (prev === "red" ? "blue" : "red"));
  };

  return (
    <div>
      <button style={buttonStyle} onClick={onClickButton} disabled={isDisabled}>
        {`change to ${buttonColor}`}
      </button>
      <div>
        <label htmlFor="disable-btn">disable button</label>
        <input
          type="checkbox"
          onChange={onClickCheckbox}
          defaultChecked={isDisabled}
          aria-checked={isDisabled}
          name="disable-btn"
          id="disable-btn"
        />
      </div>
    </div>
  );
}

export default App;
