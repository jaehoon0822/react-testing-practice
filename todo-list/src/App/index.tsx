import Button from 'components/Button';
import { Container, Contents, InputContainer, ToDoListContainer } from './App.styled';
import useApp from 'components/hooks/useApp';
import Input from 'components/Input';
import ToDoItem from 'components/ToDoItem';
import { TodoItem } from 'components/ToDoItem/ToDoItem.types';

const App = () => {
  const { onClickAddButton, onChangeTodo, todo, todoList, onDeletTodo } = useApp();
  return (
    <Container>
      <Contents>
        <ToDoListContainer data-testid="todoList">
          {todoList.map((item: TodoItem, idx: number) => {
            return (
              <ToDoItem key={item.key} label={item.label} onClickDelete={() => onDeletTodo(idx)} />
            );
          })}
        </ToDoListContainer>
        <InputContainer>
          <Input
            placeholder="할 일을 입력해 주세요"
            value={todo?.label ?? ''}
            name="todo"
            onChange={onChangeTodo}
          />
          <Button label="추가" onClick={onClickAddButton} />
        </InputContainer>
      </Contents>
    </Container>
  );
};

export default App;
