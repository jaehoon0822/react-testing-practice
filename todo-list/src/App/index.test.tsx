import { render, screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import App from './index';

jest.mock('nanoid', () => ({ nanoid: () => Math.random() }));
describe('<App />', () => {
  test('renders component correctly', async () => {
    const { container } = render(<App />);

    // toDoList 확인
    const toDoList = screen.getByTestId('todoList');
    expect(toDoList).toBeInTheDocument(); // exist

    // toDoItem 확인
    // -> toDoItem을 직접적으로 확인하기에 testid 작성하지 않음
    //    대신 '삭제' button 유무를 통해 toDoItem 존재 유무 확인
    const toDoItemButton = screen.queryByText('삭제');
    expect(toDoItemButton).toBeNull(); // null

    // input 확인
    const input = screen.getByPlaceholderText('할 일을 입력해 주세요');
    expect(input).toBeInTheDocument();

    // 추가 button 확인
    const label = screen.getByText('추가');
    expect(label).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  test('adds and deletes Todo items', async () => {
    const user = userEvent.setup();
    render(<App />);

    // input 에 study react value 작성
    const input = screen.getByPlaceholderText('할 일을 입력해 주세요');
    expect(input).toBeInTheDocument();
    await user.type(input, 'study react');

    // 작성된 input 추가
    const addButton = screen.getByText('추가');
    await user.click(addButton);

    // todoItem 확인
    const todoItem = screen.getByText('study react');
    expect(todoItem).toBeInTheDocument();

    // todoItem 삭제
    const deleteButton = screen.getByText('삭제');
    expect(deleteButton).toBeInTheDocument();
    await user.click(deleteButton);
    expect(todoItem).not.toBeInTheDocument();

    // todoItem 2개 추가
    await user.type(input, 'study react2');
    await user.click(addButton);
    await user.type(input, 'study react3');
    await user.click(addButton);

    /**** todoItem 개수 확인  ****/

    // 추가한 todoItem 존재확인
    const todoItem2 = screen.getByText('study react2');
    const todoItem3 = screen.getByText('study react3');
    expect(todoItem2).toBeInTheDocument();
    expect(todoItem3).toBeInTheDocument();

    // todoList 안의 toDoItem 갯수 확인
    const todoList = screen.getByTestId('todoList');
    const todoItems = within(todoList).getAllByTestId('toDoItem');
    expect(todoItems.length).toBe(2); // 2개 존재
  });

  test('does not add empty Todo', async () => {
    const user = userEvent.setup();
    render(<App />);

    const todoList = screen.getByTestId('todoList');
    const items = within(todoList).queryAllByTestId('toDoItem');
    const len = items.length;

    // toDoItem 이 전혀 없는 todoList 일때 toDoItme 의갯수
    expect(todoList).toBeInTheDocument();
    expect(len).toBe(0); // 추가된 item 이 없으므로 0

    // 빈 문자열일때 추가
    const addButton = screen.getByText('추가');
    await user.click(addButton);

    // 빈 문자열일때 추가되지 않아야 하므로 0
    expect(len).toBe(0);
  });
});
