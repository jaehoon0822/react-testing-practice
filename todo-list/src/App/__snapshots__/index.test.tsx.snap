// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders component correctly 1`] = `
<div>
  <div
    class="css-1q44ddc"
  >
    <div
      class="css-198ymq9"
    >
      <div
        class="css-1kumlww"
        data-testid="todoList"
      />
      <div
        class="css-ov1ktg"
      >
        <input
          class="css-1bns2mu"
          name="todo"
          placeholder="할 일을 입력해 주세요"
          type="text"
          value=""
        />
        <div
          class="css-1jzh7nd"
          data-testid="container"
        >
          <div
            class="css-68lwrs"
          >
            추가
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
